addons/controllers/cpi/cpiconfig_utils.go

```diff
@@ -0,0 +1,307 @@
+			return nil, err
+		}
+		d.VSphereCPI.Username, d.VSphereCPI.Password, err = getUsernameAndPasswordFromSecret(ctx, r.Client, vsphereSecret)
+		if err != nil {
+			return nil, err
+		}
+	}
+	d.VSphereCPI.Region = c.Region
+	d.VSphereCPI.Zone = c.Zone
+	d.VSphereCPI.InsecureFlag = c.InsecureFlag
+	d.VSphereCPI.VMInternalNetwork = c.VMInternalNetwork
+	d.VSphereCPI.VMExternalNetwork = c.VMExternalNetwork
+	d.VSphereCPI.VMExcludeInternalNetworkSubnetCidr = c.VMExcludeInternalNetworkSubnetCidr
+	d.VSphereCPI.VMExcludeExternalNetworkSubnetCidr = c.VMExcludeExternalNetworkSubnetCidr
+	d.VSphereCPI.CloudProviderExtraArgs.TLSCipherSuites = c.TLSCipherSuites
+
+	if c.NSXT != nil {
+		if c.NSXT.PodRoutingEnabled {
+			d.VSphereCPI.Nsxt.PodRoutingEnabled = c.NSXT.PodRoutingEnabled
+		}
+
+		if c.NSXT.InsecureFlag {
+			d.VSphereCPI.Nsxt.InsecureFlag = c.NSXT.InsecureFlag
+		}
+		if c.NSXT.Routes != nil {
+			d.VSphereCPI.Nsxt.Routes.RouterPath = c.NSXT.Routes.RouterPath
+			d.VSphereCPI.Nsxt.Routes.ClusterCidr = c.NSXT.Routes.ClusterCidr
+		}
+		if c.NSXT.NSXTCredentialsRef != nil {
+			nsxtSecret, err := getSecret(ctx, r.Client, c.NSXT.NSXTCredentialsRef.Namespace, c.NSXT.NSXTCredentialsRef.Name)
+			if err != nil {
+				return nil, err
+			}
+			d.VSphereCPI.Nsxt.Username, d.VSphereCPI.Nsxt.Password, err = getUsernameAndPasswordFromSecret(ctx, r.Client, nsxtSecret)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if c.NSXT.Host != "" {
+			d.VSphereCPI.Nsxt.Host = c.NSXT.Host
+		}
+		if c.NSXT.RemoteAuth {
+			d.VSphereCPI.Nsxt.RemoteAuth = c.NSXT.RemoteAuth
+		}
+		if c.NSXT.VMCAccessToken != "" {
+			d.VSphereCPI.Nsxt.VmcAccessToken = c.NSXT.VMCAccessToken
+		}
+		if c.NSXT.VMCAuthHost != "" {
+			d.VSphereCPI.Nsxt.VmcAuthHost = c.NSXT.VMCAuthHost
+		}
+		if c.NSXT.ClientCertKeyData != "" {
+			d.VSphereCPI.Nsxt.ClientCertKeyData = c.NSXT.ClientCertKeyData
+		}
+		if c.NSXT.ClientCertData != "" {
+			d.VSphereCPI.Nsxt.ClientCertData = c.NSXT.ClientCertData
+		}
+		if c.NSXT.RootCAData != "" {
+			d.VSphereCPI.Nsxt.RootCAData = c.NSXT.RootCAData
+		}
+		if c.NSXT.SecretName != "" {
+			d.VSphereCPI.Nsxt.SecretName = c.NSXT.SecretName
+			d.VSphereCPI.Nsxt.SecretNamespace = c.NSXT.SecretNamespace
+		}
+	}
+	if c.IPFamily != "" {
+		d.VSphereCPI.IPFamily = c.IPFamily
+	}
+	if c.HTTPProxy != "" {
+		d.VSphereCPI.HTTPProxy = c.HTTPProxy
+	}
+	if c.HTTPSProxy != "" {
+		d.VSphereCPI.HTTPSProxy = c.HTTPSProxy
+	}
+	if c.NoProxy != "" {
+		d.VSphereCPI.NoProxy = c.NoProxy
+	}
+	return d, nil
+}
+
+// mapCPIConfigToDataValuesParavirtual generates CPI data values for paravirtual modes
+func (r *CPIConfigReconciler) mapCPIConfigToDataValuesParavirtual(_ context.Context, cpiConfig *cpiv1alpha1.CPIConfig, _ *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	d := &CPIDataValues{}
+	c := cpiConfig.Spec.VSphereCPI
+
+	d.VSphereCPI.Mode = CPIParavirtualMode
+	d.VSphereCPI.ClusterAPIVersion = c.ClusterAPIVersion
+	d.VSphereCPI.ClusterKind = c.ClusterKind
```

Comment:
question: Can these non-paravirtual fields be derived?

Reply:
for the paravirtual mode, yes they should be derived and at least have a default value.

I would suggest do that in a separate PR since it involves other changes as well. For example, the CPIConfig controller need to generate accounts & rbac rules etc. https://gitlab.eng.vmware.com/core-build/guest-cluster-controller/-/blob/master/controllers/addons/addons_controller.go#L486-522

Reply:
A separate PR sounds ok to me. cc @vijaykatam